### PR TITLE
COMMERCE-10847 - Select order type before add product to a cart

### PR DESCRIPTION
**Pull request by ilzamcmed**
[COMMERCE-10847](https://issues.liferay.com/browse/COMMERCE-10847)

### CODE REVIEW

**Comment by ilzamcmed** [comment]
ci:test:sf

**Comment by ilzamcmed** [comment]
ci:test:relevant

**Review by interaminense — APPROVED** [review]
LGTM! Thanks @ilzamcmed

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties [hunk 1]
```diff
@@ -3680,6 +3680,7 @@ commerce-order-rule-commerce-qualifier-configuration-name=Commerce Order Rule Qu
 commerce-order-rule-entry-configuration-name=Order Rules
 commerce-order-type-commerce-qualifier-configuration-name=Commerce Order Type Qualifiers
 commerce-order-type-configuration-name=Order Types
+commerce-order-type-info=The order type cannot be modified beyond this point. If you make an error, please delete the order and begin again.
```
Comment (luca-pellizzon) [inline modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:3683]: this should only be in the "Add language key" commit, sorted properly

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties [hunk 2]
```diff
@@ -20180,4 +20180,5 @@ zip=Zip
 zip-postal=Zip/Postal
 zoom-in=Zoom In
 zoom-out=Zoom Out
-zoom-to-fit=Zoom to Fit
\ No newline at end of file
+zoom-to-fit=Zoom to Fit
+commerce-order-type-info=The order type cannot be modified beyond this point. If you make an error, please delete the order and begin again.
```
[the remark this answers is not in the document]
Comment (luca-pellizzon) [reply modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties]: If someone is doing a rebase and has conflicts with the build-lang commit, usually the commi is skipped and buildLang is re-run.
This will results in changes that are not coming directly from the user that is doing the rebase (but from your commit) and need to be manually checked. If the base key is already sorted this will not happen.

[the remark this answers is not in the document]
Comment (ilzamcmed) [reply modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties]: Got it and fixed!
Thanks @luca-pellizzon

**Comment by ilzamcmed** [comment]
I marked this PR as draft because is missing the tests updates.

**Review by luca-pellizzon — APPROVED** [review]

File: modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js [hunk 3]
```diff
@@ -115,7 +123,7 @@ describe('Add to Cart', () => {
 		});
 	});
 
-	it('must focus the quantity selector when a user tries to add to the cart an invalid quantity', async () => {
+	it.skip('must focus the quantity selector when a user tries to add to the cart an invalid quantity', async () => {
```
Comment (ilzamcmed) [inline modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js:126]: Hey @luca-pellizzon , I skiped this test because is failing and needs more investigation. I think we can move forward with this PR and I'll sent a follow up when discover whats wrong.

**Comment by ilzamcmed** [comment]
Hey @alessiorendina, can you review this PR? I accidentally removed you. :smile:

**Comment by ilzamcmed** [comment]
ci:test:sf

**Comment by ilzamcmed** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js (hunk 3 above)
Comment (luca-pellizzon) [reply modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js:126]: @victoroscuro what do you think about this?

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js [hunk 4]
```diff
@@ -58,60 +125,25 @@ function AddToCartButton({
...
+			onClick={async (event) => {
+				const CartResource = ServiceProvider.DeliveryCartAPI('v1');
+				const order = await CartResource.getCartsByAccountIdAndChannelId(
+					accountId,
+					channel.id
+				);
+
+				if (showOrderTypeModal && !order.items.length) {
+					setEvent(event);
+					Liferay.fire(OPEN_MODAL, {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:144]: Are we sure it doesn't require a prefix such as a portlet namespace? What if we have more instances of the same portlet?

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js [hunk 5]
```diff
@@ -48,6 +48,9 @@ function AddToCartButton({
 	const isMounted = useIsMounted();
 	const [event, setEvent] = useState(null);
 
+	const addToCartNotAllowed =
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:51]: Can you do a fixup for this?

File: modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js (hunk 3 above)
Comment (alessiorendina) [reply modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js:126]: Let's fix the test instead

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language_ar.properties [hunk 6]
```diff
@@ -3707,6 +3707,7 @@ commerce-order-rule-commerce-qualifier-configuration-name=مؤهلات قاعد
 commerce-order-rule-entry-configuration-name=قواعد الأمر
```
Comment (alessiorendina) [inline modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language_ar.properties:3707]: Move commits that contain auto-generated files at the end of the history

**Review by alessiorendina — CHANGES_REQUESTED** [review]
@ilzamcmed Please check my comments, thanks!

File: modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js (hunk 3 above)
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-js/test/components/AddToCart/AddToCart.js:126]: Ok, I'm working on it.

File: modules/apps/commerce/commerce-frontend-taglib/src/main/resources/META-INF/resources/add_to_cart/init.jsp [hunk 7]
```diff
@@ -50,4 +50,6 @@ String stockQuantity = (String)request.getAttribute("liferay-commerce:add-to-car
 String randomNamespace = PortalUtil.generateRandomKey(request, "taglib") + StringPool.UNDERLINE;
 
 String addToCartId = randomNamespace + "add_to_cart";
+
+boolean showOrderTypeModal = GetterUtil.getBoolean(request.getAttribute("liferay-commerce:add-to-cart:showOrderTypeModal"));
```
[the remark this answers is not in the document]
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/resources/META-INF/resources/add_to_cart/init.jsp:54]: After this change, I was getting this error ` java.lang.String cannot be cast to java.lang.Boolean`
Is there a problem if I leave `GetterUtil.getBoolean` ?

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js (hunk 5 above)
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:51]: What do you mean by fix up?

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language_ar.properties (hunk 6 above)
Comment (ilzamcmed) [reply modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language_ar.properties:3707]: I'll move it once I push the commits with the fixes you requested. Thanks

File: modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java [hunk 8]
```diff
@@ -200,6 +201,23 @@ public void setAlignment(String alignment) {
...
+			_commerceOrderTypeLocalService.getCommerceOrderTypesCount(
+				themeDisplay.getCompanyId(), true);
+
+		if (commerceOrderTypesCount > 1) {
+			setNamespacedAttribute(
+				httpServletRequest, "showOrderTypeModal", Boolean.TRUE);
+		}
+		else {
+			setNamespacedAttribute(
+				httpServletRequest, "showOrderTypeModal", Boolean.FALSE);
```
[the remark this answers is not in the document]
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java:218]: Fixed

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js (hunk 5 above)
Comment (alessiorendina) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:51]: https://git-scm.com/docs/git-commit/2.32.0#Documentation/git-commit.txt---fixupamendrewordltcommitgt

File: modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (alessiorendina) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java]: Do a fixup here as well

File: modules/apps/commerce/commerce-frontend-taglib/src/main/resources/META-INF/resources/add_to_cart/init.jsp (hunk 7 above)
[the remark this answers is not in the document]
Comment (alessiorendina) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/resources/META-INF/resources/add_to_cart/init.jsp:54]: We don't need it, we can directly cast it to primitive `boolean`

[the remark this answers is not in the document]
Comment (alessiorendina) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/resources/META-INF/resources/add_to_cart/init.jsp:54]: `(boolean)request.getAttribute("liferay-commerce:add-to-cart:showOrderTypeModal")`

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js (hunk 4 above)
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:144]: We can keep it like this because it won't open another modal, just the one to select order type

File: modules/apps/commerce/commerce-frontend-taglib/src/main/resources/META-INF/resources/add_to_cart/init.jsp (hunk 7 above)
[the remark this answers is not in the document]
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/resources/META-INF/resources/add_to_cart/init.jsp:54]: Hey @alessiorendina, I am doing like this:
`boolean showOrderTypeModal = (Boolean)request.getAttribute("liferay-commerce:add-to-cart:showOrderTypeModal");`
but keep getting the same error above. You have any idea or a better way to do it?

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language_ar.properties (hunk 6 above)
Comment (ilzamcmed) [reply modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language_ar.properties:3707]: Done

File: modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java]: Done

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js (hunk 4 above)
Comment (alessiorendina) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:144]: Can we test the behavior with 2 portlets on the same page, I have the feeling that 2 modals will be opening

File: modules/apps/commerce/commerce-frontend-taglib/src/main/resources/META-INF/resources/add_to_cart/init.jsp (hunk 7 above)
[the remark this answers is not in the document]
Comment (alessiorendina) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/resources/META-INF/resources/add_to_cart/init.jsp:54]: Let's check it together tomorrow, ping me on slack when you start your working day

**Comment by ilzamcmed** [comment]
ci:test:sf

**Comment by ilzamcmed** [comment]
ci:test:relevant

**Comment by ilzamcmed** [comment]
ci:test:relevant

**Comment by ilzamcmed** [comment]
ci:test:relevant

**Comment by luca-pellizzon** [comment]
ci:stop

**Comment by luca-pellizzon** [comment]
ci:test:relevant

**Comment by luca-pellizzon** [comment]
@ilzamcmed still tests failures in "commerce-frontend-js:packageRunTest", please check.

**Review by alessiorendina — CHANGES_REQUESTED** [review]
Hey @ilzamcmed please check my comments, thank you! I think it is better to re-do the commits from scratch at this point in order to clean the history.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js (hunk 4 above)
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:144]: Hey @alessiorendina 
I tested it and it does not open the 2 at the same time.

**Comment by ilzamcmed** [comment]
ci:test:sf

**Comment by ilzamcmed** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/views/OrdersListView.js [hunk 9]
```diff
@@ -19,6 +19,7 @@ import React, {useMemo, useRef} from 'react';
 import ServiceProvider from '../../../ServiceProvider/index';
 import {OPEN_MODAL} from '../../../utilities/eventsDefinitions';
 import {liferayNavigate} from '../../../utilities/index';
+import {MEDIUM_MODAL_SIZE} from '../../../utilities/modals/constants';
```
[the remark this answers is not in the document]
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/views/OrdersListView.js:22]: Fixed

File: modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java [hunk 10]
```diff
@@ -153,6 +153,12 @@ public int doStartTag() throws JspException {
 							 commerceContext.getCommerceSiteType()));
 				}
 			}
+
```
[the remark this answers is not in the document]
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java:157]: Fixed

**Comment by ilzamcmed** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/views/OrdersListView.js [hunk 11]
```diff
@@ -100,7 +101,10 @@ function OrdersListView({
 					displayType="primary"
 					onClick={() =>
 						showOrderTypeModal
-							? Liferay.fire(OPEN_MODAL, {id: 'add-order-modal'})
+							? Liferay.fire(OPEN_MODAL, {
```
Comment (alessiorendina) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/account_selector/views/OrdersListView.js:104]: Did you have the chance to test the behavior with 2 instances of the same portlet on the page?

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js [hunk 12]
```diff
@@ -38,10 +41,77 @@ function AddToCartButton({
...
+	useEffect(() => {
+		function handleAddNewItemToCart(orderTypeId) {
+			if (event) {
+				handleClickAddToCart(event, orderTypeId.details[0]);
+			}
+		}
+
+		Liferay.on(ADD_ITEM_TO_CART, handleAddNewItemToCart);
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
[the remark this answers is not in the document]
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:113]: Yes, because otherwise eslint gives me a warning that are unnecessary dependencies missing.

[the remark this answers is not in the document]
Comment (interaminense) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:113]: Hi Alessio, it's necessary to avoid infinite loop on useEffect

File: modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java [hunk 13]
```diff
@@ -152,6 +154,13 @@ public int doStartTag() throws JspException {
...
+			int commerceOrderTypesCount =
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java:158]: why don't you check the type of the channel?

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js [hunk 14]
```diff
@@ -58,60 +128,29 @@ function AddToCartButton({
 			disabled={disabled}
 			displayType="primary"
 			monospaced={settings.iconOnly && settings.inline}
-			onClick={(event) => {
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:61]: why do you need to move this logic?

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js [hunk 15]
```diff
@@ -58,60 +128,29 @@ function AddToCartButton({
...
-				if (cartAtomState.updating) {
-					return;
-				}
+			onClick={async (event) => {
+				const CartResource = ServiceProvider.DeliveryCartAPI('v1');
+				const order = await CartResource.getCartsByAccountIdAndChannelId(
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:133]: this looks like the problem with the guest order, why did you add this?

Comment (andrea-ale-sbarra) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:133]: this is anyway returning an array of order not just one, how do you select the right one?

**Review by andrea-ale-sbarra — CHANGES_REQUESTED** [review]

File: modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java (hunk 13 above)
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-taglib/src/main/java/com/liferay/commerce/frontend/taglib/servlet/taglib/AddToCartTag.java:158]: Because I also need to check if there is more than 1 order type to open the modal.
I added this validation on my work in progress
```
int commerceOrderTypesCount = _commerceOrderTypeLocalService.getCommerceOrderTypesCount(PortalUtil.getCompanyId(httpServletRequest), CommerceChannel.class.getName(), _commerceChannelId, true);

_showOrderTypeModal = commerceOrderTypesCount > 1 && CommerceAccountConstants.SITE_TYPE_B2B == commerceContext.getCommerceSiteType();
```

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js (hunk 15 above)
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:133]: I added this because we need to get the order count. If there is no order, I'll create a new order with the selected order type.

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js (hunk 14 above)
Comment (ilzamcmed) [reply modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/add_to_cart/AddToCartButton.js:61]: I moved this logic outside the onClick function because I need to call it to add in cart after the modal is open.

**Comment by alessiorendina** [comment]
Forwarded here: https://github.com/liferay-commerce/liferay-portal/pull/3660

Closing this one, thanks.
